examples/ex71.hpp

```diff
@@ -0,0 +1,587 @@
+#define EXAMPLE71_H
+
+#include "mfem.hpp"
+#include <memory>
+#include <iostream>
+#include <fstream>
+
+
+namespace mfem
+{
+
+
+template<typename DType, typename MVType>
+class MyQFunctorJ
+{
+public:
+   DType operator()(const mfem::Vector& vparam, MVType& uu)
+   {
+      double pp=vparam[0];
+      double ee=vparam[1];
+      double ff=vparam[2];
+
+      DType  u=uu[3];
+      DType  norm2=uu[0]*uu[0]+uu[1]*uu[1]+uu[2]*uu[2];
+
+      DType rez= pow(ee*ee+norm2,pp/2.0)/pp-ff*u;
+      return rez;
+   }
+
+   void operator()(const mfem::Vector& vparam, MVType& uu, MVType& rr)
+   {
+      double pp=vparam[0];
+      double ee=vparam[1];
+      double ff=vparam[2];
+
+      DType norm2=uu[0]*uu[0]+uu[1]*uu[1]+uu[2]*uu[2];
+      DType tvar=pow(ee*ee+norm2,(pp-2.0)/2.0);
+
+      rr[0]=tvar*uu[0];
+      rr[1]=tvar*uu[1];
+      rr[2]=tvar*uu[2];
+      rr[3]=-ff;
+
+   }
+};
+
+
+typedef ADQFunctionTJ<MyQFunctorJ,4> pLapIntegrandTJ;
+
+
+template<typename DType, typename MVType>
+class MyQFunctorH
+{
+public:
+   DType operator()(const mfem::Vector& vparam, MVType& uu)
+   {
+      double pp=vparam[0];
+      double ee=vparam[1];
+      double ff=vparam[2];
+
+      DType  u=uu[3];
+      DType  norm2=uu[0]*uu[0]+uu[1]*uu[1]+uu[2]*uu[2];
+
+      DType rez= pow(ee*ee+norm2,pp/2.0)/pp-ff*u;
+      return rez;
+   }
+};
+
+typedef ADQFunctionTH<MyQFunctorH> pLapIntegrandTH;
+
+//comment the line below in order to use
+//pLapIntegrandTJ for differentiation
+//the user interface for both TH and TJ versions
+//is exacly the same
+//#define USE_ADH
+
+class pLaplaceAD: public mfem::NonlinearFormIntegrator
+{
+protected:
+   mfem::Coefficient* pp;
+   mfem::Coefficient* coeff;
+   mfem::Coefficient* load;
+#ifdef USE_ADH
+   pLapIntegrandTH qint;
+#else
+   pLapIntegrandTJ qint;
+#endif
+public:
+   pLaplaceAD()
+   {
+      coeff=nullptr;
+      pp=nullptr;
+   }
+
+   pLaplaceAD(mfem::Coefficient& pp_):pp(&pp_), coeff(nullptr), load(nullptr)
+   {
+
+   }
+
+   pLaplaceAD(mfem::Coefficient &pp_,mfem::Coefficient& q,
+              mfem::Coefficient& ld_): pp(&pp_), coeff(&q), load(&ld_)
+   {
+
+   }
+
+   virtual ~pLaplaceAD()
+   {
+
+   }
+
+   virtual double GetElementEnergy(const mfem::FiniteElement &el,
+                                   mfem::ElementTransformation &trans, const mfem::Vector &elfun) override
+   {
+      double energy=0.0;
+      int ndof = el.GetDof();
+      int ndim = el.GetDim();
+      int spaceDim = trans.GetSpaceDim();
+      bool square = (ndim == spaceDim);
+      const mfem::IntegrationRule *ir = NULL;
+      int order = 2 * trans.OrderGrad(&el) - 1; // correct order?
+      ir = &mfem::IntRules.Get(el.GetGeomType(), order);
+
+      mfem::Vector shapef(ndof);
+      mfem::DenseMatrix dshape_iso(ndof,ndim);
+      mfem::DenseMatrix dshape_xyz(ndof,spaceDim);
+      mfem::Vector grad(spaceDim);
+
+      mfem::Vector vparam(3);//[power, epsilon, load]
+      mfem::Vector uu(4);//[diff_x,diff_y,diff_z,u]
```

Comment:
Special indices should be captured in enums.

Reply:
Again - these indices are case dependent - I am not sure enum is the correct solution here as it will have to be defined for every new implementation.